lib/block_view_stake.go

```diff
@@ -0,0 +1,2496 @@
+		return RuleErrorInvalidStakeInsufficientBalance
+	}
+
+	return nil
+}
+
+func (bav *UtxoView) IsValidUnstakeMetadata(transactorPkBytes []byte, metadata *UnstakeMetadata) error {
+	// Validate TransactorPublicKey.
+	transactorPKIDEntry := bav.GetPKIDForPublicKey(transactorPkBytes)
+	if transactorPKIDEntry == nil || transactorPKIDEntry.isDeleted {
+		return RuleErrorInvalidStakerPKID
+	}
+
+	// Validate ValidatorPublicKey.
+	validatorEntry, err := bav.GetValidatorByPublicKey(metadata.ValidatorPublicKey)
+	if err != nil {
+		return errors.Wrapf(err, "IsValidUnstakeMetadata: ")
+	}
+	if validatorEntry == nil || validatorEntry.isDeleted {
+		return RuleErrorInvalidValidatorPKID
+	}
+
+	// Validate StakeEntry exists.
+	stakeEntry, err := bav.GetStakeEntry(validatorEntry.ValidatorPKID, transactorPKIDEntry.PKID)
+	if err != nil {
+		return errors.Wrapf(err, "IsValidUnstakeMetadata: ")
+	}
+	if stakeEntry == nil || stakeEntry.isDeleted {
+		return RuleErrorInvalidUnstakeNoStakeFound
+	}
+
+	// Validate 0 < UnstakeAmountNanos <= StakeEntry.StakeAmountNanos.
+	if metadata.UnstakeAmountNanos == nil || metadata.UnstakeAmountNanos.IsZero() {
+		return RuleErrorInvalidUnstakeAmountNanos
+	}
+	if stakeEntry.StakeAmountNanos.Cmp(metadata.UnstakeAmountNanos) < 0 {
+		return RuleErrorInvalidUnstakeInsufficientStakeFound
+	}
+
+	return nil
+}
+
+func (bav *UtxoView) IsValidUnlockStakeMetadata(transactorPkBytes []byte, metadata *UnlockStakeMetadata) error {
+	// Validate TransactorPublicKey.
+	transactorPKIDEntry := bav.GetPKIDForPublicKey(transactorPkBytes)
+	if transactorPKIDEntry == nil || transactorPKIDEntry.isDeleted {
+		return RuleErrorInvalidStakerPKID
+	}
+
+	// Validate ValidatorPublicKey.
+	validatorEntry, err := bav.GetValidatorByPublicKey(metadata.ValidatorPublicKey)
+	if err != nil {
+		return errors.Wrapf(err, "IsValidUnlockStakeMetadata: ")
+	}
+	if validatorEntry == nil || validatorEntry.isDeleted {
+		return RuleErrorInvalidValidatorPKID
+	}
+
+	// Validate StartEpochNumber and EndEpochNumber.
+	if metadata.StartEpochNumber > metadata.EndEpochNumber {
+		return RuleErrorInvalidUnlockStakeEpochRange
+	}
+	// TODO: validate EndEpochNumber is <= CurrentEpochNumber - 2
+
+	// Validate LockedStakeEntries exist.
+	lockedStakeEntries, err := bav.GetLockedStakeEntriesInRange(
+		validatorEntry.ValidatorPKID, transactorPKIDEntry.PKID, metadata.StartEpochNumber, metadata.EndEpochNumber,
+	)
+	existsLockedStakeEntries := false
+	for _, lockedStakeEntry := range lockedStakeEntries {
+		if lockedStakeEntry != nil && !lockedStakeEntry.isDeleted {
+			existsLockedStakeEntries = true
+			break
+		}
+	}
+	if !existsLockedStakeEntries {
+		return RuleErrorInvalidUnlockStakeNoUnlockableStakeFound
+	}
+
+	return nil
+}
+
+func (bav *UtxoView) GetStakeEntry(validatorPKID *PKID, stakerPKID *PKID) (*StakeEntry, error) {
+	// First, check the UtxoView.
+	stakeMapKey := StakeMapKey{ValidatorPKID: *validatorPKID, StakerPKID: *stakerPKID}
```

Comment:
do we want to check that validator PKID and staker PKID is not nil before we do this? just prevents a panic.